sql-plugin/src/main/scala/org/apache/spark/sql/rapids/execution/GpuHashJoin.scala

```diff
@@ -622,7 +628,114 @@ trait GpuHashJoin extends GpuExec {
     (joinLeft.output.size, boundCondition)
   }
 
-  def doJoin(
+  private def existenceJoinIterator(
+    builtBatch: ColumnarBatch,
+    stream: Iterator[ColumnarBatch]) = new Iterator[ColumnarBatch]()
+      with AutoCloseable with Arm {
+    var closed: Boolean = false
+
+    // iteration-independent resources
+    val resources = ArrayBuffer[AutoCloseable]()
+    def use[T <: AutoCloseable](ac: T): T = {
+      resources += ac
+      ac
+    }
+
+    val compiledConditionRes: Option[(Table, CompiledExpression)] = boundCondition.map(gpuExpr => (
+        use(GpuColumnVector.from(builtBatch)),
+        use(gpuExpr.convertToAst(numFirstConditionTableColumns).compile())
+      )
+    )
+
+    val rightKeysTab = use(
+      withResource(GpuProjectExec.project(builtBatch, boundBuildKeys))(GpuColumnVector.from(_)))
+
+    val falseScalar = use(Scalar.fromBool(false))
+    val trueScalar = use(Scalar.fromBool(true))
```

Comment:
they are closed when the iterator is exhausted. Why keep recreating them?

Reply:
Because they can create fragmentation problems by holding onto them for a long time.  The counter argument is: they are relatively cheap to create, so why cache them?